thewalrus/tests/test_internal_modes.py

```diff
@@ -0,0 +1,1098 @@
+    hbar = 2
+    rjs = [np.array([r]), np.array([r])]
+    O = np.array([[1, S * np.exp(-1j * phi)], [S * np.exp(1j * phi), 1]])
+    U = unitary_group.rvs(len(rjs))
+    Q = prepare_cov(rjs, U, O=O, thresh=0.0, hbar=hbar)
+    W0 = np.array([[np.sqrt(1 + S), np.sqrt(1 - S)], [np.sqrt(1 - S), -np.sqrt(1 + S)]]) / np.sqrt(
+        2
+    )
+    W1 = (
+        np.array([[np.sqrt(1 + S), -np.sqrt(1 - S)], [np.sqrt(1 - S), np.sqrt(1 + S)]])
+        * np.exp(-1j * phi)
+        / np.sqrt(2)
+    )
+    eps, W = [np.array([r, 0]), np.array([r, 0])], [W0, W1]
+    Qinit = (hbar / 2) * np.diag(
+        np.array([np.exp(-2 * r), 1, np.exp(-2 * r), 1, np.exp(2 * r), 1, np.exp(2 * r), 1])
+    )
+    Uw = np.block([[W0.T.conj(), np.zeros(W0.shape)], [np.zeros(W1.shape), W1.T.conj()]])
+    Qorth = interferometer(Uw) @ Qinit @ interferometer(Uw).T
+    Qu = implement_U(Qorth, U)
+    assert np.allclose(Q, Qu)
+
+
+def test_pure_gkp():
+    """test pure gkp state density matrix using 2 methods from the walrus against
+    internal_modes.density_matrix_single_mode (but with only 1 temporal mode)"""
+
+    m1, m2 = 5, 7
+    params = np.array(
+        [
+            -1.38155106,
+            -1.21699567,
+            0.7798817,
+            1.04182349,
+            0.87702211,
+            0.90243916,
+            1.48353639,
+            1.6962906,
+            -0.24251599,
+            0.1958,
+        ]
+    )
+    sq_r = params[:3]
+    bs_theta1, bs_theta2, bs_theta3 = params[3:6]
+    bs_phi1, bs_phi2, bs_phi3 = params[6:9]
+    sq_virt = params[9]
+
+    S1 = squeezing(np.abs(sq_r), phi=np.angle(sq_r))
+    U1 = np.array(
+        [
+            [np.cos(bs_theta1), -np.exp(-1j * bs_phi1) * np.sin(bs_theta1), 0],
+            [np.exp(1j * bs_phi1) * np.sin(bs_theta1), np.cos(bs_theta1), 0],
+            [0, 0, 1],
+        ]
+    )
+    U2 = np.array(
+        [
+            [1, 0, 0],
+            [0, np.cos(bs_theta2), -np.exp(-1j * bs_phi2) * np.sin(bs_theta2)],
+            [0, np.exp(1j * bs_phi2) * np.sin(bs_theta2), np.cos(bs_theta2)],
+        ]
+    )
+    U3 = np.array(
+        [
+            [np.cos(bs_theta3), -np.exp(-1j * bs_phi3) * np.sin(bs_theta3), 0],
+            [np.exp(1j * bs_phi3) * np.sin(bs_theta3), np.cos(bs_theta3), 0],
+            [0, 0, 1],
+        ]
+    )
+    Usymp = interferometer(U3 @ U2 @ U1)
+    r2 = np.array([0, 0, sq_virt])
+    S2 = squeezing(np.abs(r2), phi=np.angle(r2))
+    Z = S2 @ Usymp @ S1
+    cov = Z @ Z.T
+
+    cutoff = 26
+    mu = np.zeros([len(cov)])
+
+    psi = state_vector(mu, cov, post_select={1: m1, 2: m2}, normalize=False, cutoff=cutoff)
+
+    rho1 = np.outer(psi, psi.conj())
+    rho1 /= np.trace(rho1)
+    # get density matrix directly using The Walrus
+    rho2 = density_matrix(mu, cov, post_select={1: m1, 2: m2}, cutoff=cutoff)
+    rho2 /= np.trace(rho2)
+
+    # get density matrix using new code
+    rho3 = density_matrix_single_mode(cov, {1: m1, 2: m2}, cutoff=cutoff - 1)
```

Comment:
The Fock support of this output state should only be on even components. In rho1, the coefficients for the odd components are always 0. In rho2, the odd Fock support is at most order 1e-9. In rho3, some odd terms are as high as 1e-4, which seems to be the source of why @nquesada had to raise the tolerances to be so high.